### PR TITLE
Refactor semantic text field to align with text field behaviour  

### PR DESCRIPTION
**Pull request by jimczi**
This PR updates the semantic text field to function similarly to a standard text field. The original source content is preserved, while embedding chunks are retrieved from:  
- Doc values for dense embeddings  
- Term vectors for sparse embeddings  

The new format is not enabled by default. An internal index setting is used to enable the new format, allowing for a controlled transition. The switch to the new format will be implemented in a subsequent PR.  
Highlighting support is also missing (for the new format) and will be added in a follow up.

Note: Individual commits were reviewed in the original branch for clarity. 
This PR is intended for merging rather than detailed review.

### CODE REVIEW

**Review by Mikep86 — COMMENTED** [review]
Partial review, but I think the comments cover the main points to discuss.

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java [hunk 1]
```diff
@@ -253,14 +282,14 @@ public void testInvalidInferenceEndpoints() {
         {
             Exception e = expectThrows(
                 MapperParsingException.class,
-                () -> createMapperService(fieldMapping(b -> b.field("type", "semantic_text").field(INFERENCE_ID_FIELD, "")))
+                () -> createMapperService(fieldMapping(b -> b.field("type", "semantic_text").field(INFERENCE_ID_FIELD, "")), false)
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java:285]: We should use `useLegacyFormat` here

**Review by Mikep86 — APPROVED** [review]
LGTM! I flagged some things we should address in follow-ups.

File: docs/changelog/119183.yaml
[the remark this answers is not in the document]
Comment (jimczi) [reply docs/changelog/119183.yaml:1]: Addressed in https://github.com/elastic/elasticsearch/pull/119183/commits/05458c6b661e2ae2b172e17b62e906f993cfbd26

File: server/src/main/java/org/elasticsearch/index/engine/TranslogDirectoryReader.java [hunk 2]
```diff
@@ -440,7 +440,7 @@ private void readStoredFieldsDirectly(StoredFieldVisitor visitor) throws IOExcep
                 SourceFieldMapper mapper = mappingLookup.getMapping().getMetadataMapperByClass(SourceFieldMapper.class);
                 if (mapper != null) {
                     try {
-                        sourceBytes = mapper.applyFilters(sourceBytes, null);
+                        sourceBytes = mapper.applyFilters(null, sourceBytes, null);
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/engine/TranslogDirectoryReader.java]: Addressed in https://github.com/elastic/elasticsearch/pull/119183/commits/05458c6b661e2ae2b172e17b62e906f993cfbd26

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticInferenceMetadataFieldsMapper.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+        @Override
+        public ValueFetcher valueFetcher(MappingLookup mappingLookup, Function<Query, BitSetProducer> bitSetCache, IndexSearcher searcher) {
+            Map<String, ValueFetcher> fieldFetchers = new HashMap<>();
+            for (var inferenceField : mappingLookup.inferenceFields().keySet()) {
+                MappedFieldType ft = mappingLookup.getFieldType(inferenceField);
+                if (ft instanceof SemanticTextFieldMapper.SemanticTextFieldType semanticTextFieldType) {
+                    fieldFetchers.put(inferenceField, semanticTextFieldType.valueFetcherWithInferenceResults(bitSetCache, searcher));
+                } else {
+                    throw new IllegalArgumentException("Illegal format for field [" + name() + "], got " + ft.typeName());
```
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticInferenceMetadataFieldsMapper.java]: Addressed in https://github.com/elastic/elasticsearch/pull/119183/commits/05458c6b661e2ae2b172e17b62e906f993cfbd26

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java [hunk 4]
```diff
@@ -253,14 +282,14 @@ public void testInvalidInferenceEndpoints() {
...
             );
             assertThat(e.getMessage(), containsString("[inference_id] on mapper [field] of type [semantic_text] must not be empty"));
         }
         {
             Exception e = expectThrows(
                 MapperParsingException.class,
-                () -> createMapperService(fieldMapping(b -> b.field("type", "semantic_text").field(SEARCH_INFERENCE_ID_FIELD, "")))
+                () -> createMapperService(fieldMapping(b -> b.field("type", "semantic_text").field(SEARCH_INFERENCE_ID_FIELD, "")), false)
```
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java]: Done in https://github.com/elastic/elasticsearch/pull/119183/commits/05458c6b661e2ae2b172e17b62e906f993cfbd26

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java [hunk 5]
```diff
@@ -471,12 +502,19 @@ private static void assertSemanticTextField(MapperService mapperService, String
...
+
+        Mapper textMapper = chunksMapper.getMapper(TEXT_FIELD);
+        if (semanticTextFieldType.useLegacyFormat()) {
+            assertNotNull(textMapper);
+            assertThat(textMapper, instanceOf(KeywordFieldMapper.class));
+            KeywordFieldMapper textFieldMapper = (KeywordFieldMapper) textMapper;
+            assertFalse(textFieldMapper.fieldType().isIndexed());
+            assertFalse(textFieldMapper.fieldType().hasDocValues());
+        } else {
+            // TODO: Check for offsets mapper here
```
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java]: Done in https://github.com/elastic/elasticsearch/pull/119183/commits/05458c6b661e2ae2b172e17b62e906f993cfbd26

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java]: Done in https://github.com/elastic/elasticsearch/pull/119183/commits/05458c6b661e2ae2b172e17b62e906f993cfbd26

File: server/src/main/java/org/elasticsearch/action/update/UpdateHelper.java [hunk 6]
```diff
@@ -60,7 +60,15 @@ public UpdateHelper(ScriptService scriptService) {
...
-        final GetResult getResult = indexShard.getService().getForUpdate(request.id(), request.ifSeqNo(), request.ifPrimaryTerm());
+        // TODO: Don't hard-code gFields
+        return prepare(request, indexShard, nowInMillis, new String[] { RoutingFieldMapper.NAME });
+    }
+
+    /**
+     * Prepares an update request by converting it into an index or delete request or an update response (no action).
+     */
+    public Result prepare(UpdateRequest request, IndexShard indexShard, LongSupplier nowInMillis, String[] gFields) throws IOException {
+        final GetResult getResult = indexShard.getService().getForUpdate(request.id(), request.ifSeqNo(), request.ifPrimaryTerm(), gFields);
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/update/UpdateHelper.java:71]: The `UpdateHelper` is used by many threads so we cannot "set" fields there.

File: server/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 7]
```diff
@@ -372,6 +387,19 @@ private GetResult innerGetFetch(
...
+                source = source.filter(filter);
+            }
+
+            if (hasInferenceMetadataFields) {
+                /**
+                 * Adds the {@link InferenceMetadataFieldsMapper#NAME} field from the document fields
+                 * to the original _source if it has been requested.
+                 */
+                source = addInferenceMetadataFields(mapperService, docIdAndVersion.reader.getContext(), docIdAndVersion.docId, source);
+            }
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/get/ShardGetService.java:402]: I think it's fine, it's a metadata field that is added in _source on demand.

**Pull request merged**